src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+  dout(20) << "oid=" << frag_oid.name
+           << " dname=" << dname
+           << " frag=" << frag
+           << ", r=" << r << dendl;
+  if (r < 0) {
+    return r;
+  }
+
+  if (vals.find(key) == vals.end()) {
+    dout(20) << key << " not found in result" << dendl;
+    return -ENOENT;
+  }
+
+  try {
+    bufferlist::iterator q = vals[key].begin();
+    snapid_t dnfirst;
+    ::decode(dnfirst, q);
+    char dentry_type;
+    ::decode(dentry_type, q);
+    if (dentry_type == 'I') {
+      inode->decode_bare(q);
+      return 0;
+    } else {
+      dout(20) << "dentry type '" << dentry_type << "': cannot"
+                  "read an inode out of that" << dendl;
+      return -EINVAL;
+    }
+  } catch (const buffer::error &err) {
+    dout(20) << "encoding error in dentry 0x" << std::hex << parent_ino
+             << std::dec << "/" << dname << dendl;
+    return -EINVAL;
+  }
+
+  return 0;
+}
+
+int MetadataDriver::inject_lost_and_found(
+    inodeno_t ino, uint64_t file_size, time_t file_mtime, uint32_t chunk_size,
+    int64_t data_pool_id)
+{
+  // Create lost+found if doesn't exist
+  bool created = false;
+  int r = find_or_create_dirfrag(CEPH_INO_ROOT, frag_t(), &created);
+  if (r < 0) {
+    return r;
+  }
+  InodeStore lf_ino;
+  r = read_dentry(CEPH_INO_ROOT, frag_t(), "lost+found", &lf_ino);
+  if (r == -ENOENT || r == -EINVAL) {
+    // Inject dentry
+    lf_ino.inode.mode = 0755 | S_IFDIR;
+    lf_ino.inode.dirstat.nfiles = 1;
+    lf_ino.inode.size = 1;
+    lf_ino.inode.nlink = 1;
+    lf_ino.inode.ino = CEPH_INO_LOST_AND_FOUND;
+    lf_ino.inode.version = 1;
+    lf_ino.inode.backtrace_version = 1;
+    // TODO: we should probably fragment this by default, as we will
+    // potentially write very many dentries.
+    r = inject_linkage(CEPH_INO_ROOT, "lost+found", frag_t(), lf_ino);
+    if (r < 0) {
+      return r;
+    }
+  } else {
+    if (!(lf_ino.inode.mode & S_IFDIR)) {
+      derr << "lost+found exists but is not a directory!" << dendl;
+      // In this case we error out, and the user should do something about
+      // this problem.
+      return -EINVAL;
+    }
+  }
+
+  r = find_or_create_dirfrag(CEPH_INO_LOST_AND_FOUND, frag_t(), &created);
+  if (r < 0) {
+    return r;
+  }
+
+  InodeStore recovered_ino;
+  recovered_ino.inode.mode = 0500 | S_IFREG;
+  recovered_ino.inode.size = file_size;
+  recovered_ino.inode.max_size_ever = file_size;
+  recovered_ino.inode.mtime.tv.tv_sec = file_mtime;
+  recovered_ino.inode.atime.tv.tv_sec = file_mtime;
+  recovered_ino.inode.ctime.tv.tv_sec = file_mtime;
+
+  recovered_ino.inode.layout = g_default_file_layout;
+  recovered_ino.inode.layout.fl_object_size = chunk_size;
+  recovered_ino.inode.layout.fl_stripe_unit = chunk_size;
+  recovered_ino.inode.layout.fl_pg_pool = data_pool_id;
+
+  recovered_ino.inode.truncate_seq = 1;
+  recovered_ino.inode.truncate_size = -1ull;
+
+  recovered_ino.inode.inline_data.version = CEPH_INLINE_NONE;
```

Comment:
Doing nothing is pretty much all we can do when recovering from non-inlined on-disk data -- if metadata lost, no way of knowing that something was ever inline